modules/caddyhttp/reverseproxy/upstreams.go

```diff
@@ -140,6 +147,12 @@ func (su SRVUpstreams) GetUpstreams(r *http.Request) ([]*Upstream, error) {
 		// out and an error will be returned alongside the remaining results, if any." Thus, we
 		// only return an error if no records were also returned.
 		if len(records) == 0 {
+			if su.GracePeriod > 0 {
+				su.logger.Error("SRV lookup failed; using previously cached", zap.Error(err))
+				cached.freshness = time.Now().Add(-time.Duration(su.Refresh) - time.Duration(su.GracePeriod))
```

Comment:
I'm second-guessing my math here.

I should write some tests for this. I did a couple in my head but I should probably just write them down.

Reply:
Still second guessing, or should we merge?